crates/subspace-networking/src/utils/piece_provider.rs

```diff
@@ -90,18 +90,15 @@ where
     /// Get pieces with provided indices from cache.
     ///
     /// Number of elements in returned stream is the same as number of unique `piece_indices`.
-    pub async fn get_from_cache<'a, PieceIndices>(
+    pub async fn get_from_cache<'a>(
         &'a self,
-        piece_indices: PieceIndices,
-    ) -> impl Stream<Item = (PieceIndex, Option<Piece>)> + Unpin + 'a
-    where
-        PieceIndices: IntoIterator<Item = PieceIndex> + 'a,
-    {
+        piece_indices: Vec<PieceIndex>,
```

Comment:
Commit message says avoid allocation, but this change forces allocation on API users

Reply:
I tried removing the allocations, and the ergonomics are substantially worse. It also reached a point where the compiler couldn't infer the types of some trait method generics (and there was a compiler message saying that), so they all had to be specified manually.

I think this is the best we can do on current compiler versions.

Reply:
> current API users have already allocated a vector

That was the whole point though, wasn't it? If the caller needs to allocate - they do, but API didn't require it. Now the API requires allocation no matter what.

I guess I don't fully understand why did it need to change, or rather why removing allocation in one place requires allocation in another and how that is an improvement overall.

Reply:
Unfortunately that won't work, because some of the iterators are non-contiguous (and `AsRef<[PieceIndex]>` can only operate on contiguous memory without allocating an array or vector).

And trying to replace `Vec<PieceIndex>` in the `PieceGetter` trait runs into dyn compatibility and lifetime issues.

Maybe this is the best we can do for now.

Reply:
Just following up on this, it appears to be a known Rust compiler issue with incorrect lifetime bounds in async:
https://github.com/rust-lang/rust/issues/110338

Reply:
No, this is not related to async. We just iterate over the collection we're also trying to modify.

Reply:
I was talking about the PieceGetter async trait, and why we have to pass it a vector, rather than a generic IntoIterator type.